### PR TITLE
Update composer and tests to support php 8

### PR DESCRIPTION
**Pull request by sunkan**
This is the smallest change I could make and make the tests pass and work on both php 7.1 and php 8.0.0-rc1

I haven't update the `ci.yml` due to not being familiar with that setup

### CODE REVIEW

**Comment by chris-doehring** [comment]
Hi @sunkan,

I was also trying to make Buzz php 8 compatible a few days ago but ran into an incompatibility regarding the multi curl server push functionality. There must be some bug or change in the curl version that php 8 is using, that makes the multi curl server push support of this package incompatible. I deactivated it for php 8 as it is  only activated for php 7.4 anyway.

You can check out the progress I did so far: https://github.com/kriswallsmith/Buzz/compare/master...chris-doehring:feature/php8

Maybe the best thing would be to combine efforts here? I think I will create a pull request in your fork to add the remaining changes I did including the `ci.yml` changes, so it will be in this pull request after merging.

**Comment by sunkan** [comment]
@Nyholm 

This now passes all checks thanks to @chris-doehring 

Could we get this merged?

File: composer.json [hunk 1]
```diff
@@ -48,5 +48,8 @@
         "psr-4": {
             "Buzz\\Test\\": "tests"
         }
+    },
+    "conflict": {
+        "symfony/polyfill-php80": "*"
```
Comment (Nyholm) [inline composer.json:53]: Why is this needed?

File: lib/Client/AbstractClient.php [hunk 2]
```diff
@@ -13,7 +13,7 @@
 abstract class AbstractClient
 {
     /**
-     * @var OptionsResolver
+     * @var OptionsResolver|null
```
Comment (Nyholm) [inline lib/Client/AbstractClient.php:16]: Is this really true?

File: lib/Client/MultiCurl.php [hunk 3]
```diff
@@ -56,7 +56,13 @@ public function __construct($responseFactory, array $options = [])
     {
         parent::__construct($responseFactory, $options);
 
-        if (\PHP_VERSION_ID < 70215 || \PHP_VERSION_ID === 70300 || \PHP_VERSION_ID === 70301 || !(CURL_VERSION_HTTP2 & curl_version()['features'])) {
+        if (
+            \PHP_VERSION_ID < 70215 ||
+            \PHP_VERSION_ID === 70300 ||
+            \PHP_VERSION_ID === 70301 ||
+            \PHP_VERSION_ID >= 80000 ||
```
Comment (Nyholm) [inline lib/Client/MultiCurl.php:63]: Why disable h2 push on php8?

File: tests/Integration/MultiCurlServerPushTest.php [hunk 4]
```diff
@@ -10,10 +10,10 @@
 
 class MultiCurlServerPushTest extends BaseIntegrationTest
 {
-    protected function setUp()
+    protected function setUp(): void
     {
         parent::setUp();
-        if (\PHP_VERSION_ID < 70400 || curl_version()['version_number'] < 472065) {
+        if (\PHP_VERSION_ID < 70400 || \PHP_VERSION_ID >= 80000 || curl_version()['version_number'] < 472065) {
```
Comment (Nyholm) [inline tests/Integration/MultiCurlServerPushTest.php:16]: Same

**Review by Nyholm — CHANGES_REQUESTED** [review]
Thank you for the ping. 

The pr looks good. I have a few questions I don’t understand.

File: composer.json (hunk 1 above)
Comment (sunkan) [reply composer.json:53]: From @chris-doehring : 
due to the combination of different symfony packages, I had to set the package symfony/polyfill-php80 as conflict so it will be ignored for installation (credits: laravel/framework#33045 (comment))

File: lib/Client/AbstractClient.php (hunk 2 above)
Comment (sunkan) [reply lib/Client/AbstractClient.php:16]: Theoretically it's correct since it's populated in the getter and you do a null check in it on line 46

File: lib/Client/MultiCurl.php (hunk 3 above)
Comment (sunkan) [reply lib/Client/MultiCurl.php:63]: From @chris-doehring :

due to some strange curl behaviour under php 8, the multi curl server push mechanism had to be disabled for php 8. That's something someone needs to check in another pull request.

I was gonna leave the problems with push for you to fix.

Comment (chris-doehring) [reply lib/Client/MultiCurl.php:63]: While I was testing my changes, the `MultiCurlServerPushTest` always failed under php 8, as the `$timeFirstRequest > $timeOtherRequests` assertion was false. I looked into the curl changelog, since it works under php 7.4 but found nothing significant which would make the implementation fail, but either the support in this package, something new of php 8 or in curl is responsible for this. In favor of this PR it was the best in my opinion to deactivate it for php 8 and create a new issue for this so someone can take a deeper dive into it.

Comment (chris-doehring) [reply lib/Client/MultiCurl.php:63]: At least one issue is this weak comparison in the MultiCurl class, since resource comparing changed in php 8, but it wasn't the only issue issue responsible for making the test fail

image

File: composer.json (hunk 1 above)
Comment (Nyholm) [reply composer.json:53]: I still dont understand why it should be ignored for installation.

File: lib/Client/AbstractClient.php (hunk 2 above)
Comment (Nyholm) [reply lib/Client/AbstractClient.php:16]: Yes, you are correct. 

Thank you

File: composer.json (hunk 1 above)
Comment (chris-doehring) [reply composer.json:53]: I will try to reproduce the error I got while running one of the CI pipelines.

File: lib/Client/MultiCurl.php (hunk 3 above)
Comment (Nyholm) [reply lib/Client/MultiCurl.php:63]: Okey. Let's create a new issue about h2 push in php8 and merge this as it is.

File: composer.json (hunk 1 above)
Comment (chris-doehring) [reply composer.json]: I don't have the time to do it now, but will check it out this evening

Comment (sunkan) [reply composer.json]: I think the problem is that we run `composer install` on php7.4 
but the tests run on php8-dev from `tommymuehle/docker-alpine-php-nightly`

https://github.com/kriswallsmith/Buzz/blob/7b43248c5bad1eb8db805cd1fb8ed77ee891ffc7/.github/workflows/ci.yml#L102-L125

Comment (chris-doehring) [reply composer.json]: Good catch! I assumed it was still based on php 7.4, as it's docker hub readme suggested it: https://hub.docker.com/r/tommymuehle/docker-alpine-php-nightly/

**Review by Nyholm — APPROVED** [review]
Cool cool cool. 

Thank you
